mlflow/utils/async_logging/async_logging_queue.py

```diff
@@ -0,0 +1,161 @@
+        self._continue_to_log_data = threading.Event()
+        # Keeping max_workers=1 so that there are no two threads
+        self._BATCH_LOGGING_THREADPOOL = ThreadPoolExecutor(max_workers=1)
+
+        self._BATCH_STATUS_CHECK_THREADPOOL = ThreadPoolExecutor(max_workers=10)
+
+        self._run_data_logging_thread = self._BATCH_LOGGING_THREADPOOL.submit(
+            self._logging_loop
+        )  # concurrent.futures.Future[self._logging_loop]
+
+        atexit.register(self._at_exit_callback)
+
+    def _at_exit_callback(self) -> None:
+        """
+        Callback function to be executed when the program is exiting.
+        Stops the data processing thread and waits for the
+        queue to be drained.
+        Finally, shuts down the thread pools used for data logging and batch processing status
+          check.
+        """
+        try:
+            # Stop the data processing thread
+            self._continue_to_log_data.set()
+            # Waits till queue is drained.
+            self._run_data_logging_thread.result()
+            self._BATCH_LOGGING_THREADPOOL.shutdown(wait=False)
+            self._BATCH_STATUS_CHECK_THREADPOOL.shutdown(wait=False)
+        except Exception as e:
+            _logger.error(f"Encountered error while trying to finish logging: {e}")
+
+    def _logging_loop(self) -> None:
+        """
+        Continuously logs run data until `self._continue_to_process_data` is set to False.
+        If an exception occurs during logging, a `MlflowException` is raised.
+        """
+        try:
+            while not self._continue_to_log_data.is_set():
+                self._log_run_data()
+        except Exception as e:
+            from mlflow.exceptions import MlflowException
+
+            raise MlflowException(f"Exception inside the run data logging thread: {e}")
+
+    def _log_run_data(self) -> None:
+        """
+        Process the run data in the running runs queues.
+
+        For each run in the running runs queues, this method retrieves the next batch of run data
+         from the queue and processes it by calling the `_processing_func` method with the run ID,
+           metrics, parameters, and tags in the batch.
+        If the batch is empty, it is skipped. After processing the batch, the processed watermark
+         is updated and the batch event is set.
+        If an exception occurs during processing, the exception is logged and the batch event is set
+        with the exception. If the queue is empty, it is ignored.
+
+        Returns: None
+        """
+        run_batch = None  # type: RunBatch
+        try:
+            run_batch = self._queue.get(timeout=1)
+        except Empty:
+            # Ignore empty queue exception
+            return
+        try:
+            self._logging_func(
+                run_id=run_batch.run_id,
+                metrics=run_batch.metrics,
+                params=run_batch.params,
+                tags=run_batch.tags,
+            )
```

Comment:
@sagarsumant @chenmoneygithub reopening the previous discussion about batching. Here's the problem:

In recent versions of MLflow, if I call `log_params()` with 10,000 params, MLflow will divide the 10,000 params into batches of 100 and log each batch sequentially to avoid errors (e.g. the REST API on Databricks only allows 100 params to be logged per request).

With this PR, if I pass 10,000 params with `synchronous=False`, the implementation will try to log all 10,000 params in one API request and fail.

Can we fix this and add a test case?

Reply:
Did some manual testing myself, and it looks like `synchronous=False` doesn't try to log all 10,000 params in one REST API request. @sagarsumant can you confirm?

Reply:
Edit: This causes problems with `wait()`. If I pass in 10,000 params and then call `wait()`, it returns almost immediately, even though the params haven't been logged. Can we fix this?

Reply:
Good catch, I think the code is expecting `wait()` to wait until all params are logged, basically the `RunOperations` is a combined result. Maybe there is some bug, let's add a test covering the case.

Reply:
Thanks for reporting this, will verify and add a test case for the same.

Reply:
Verified, fixed and added test case. Thanks.